File: IntegrationTests/Parser/TypeParsing/TypeParserTests/Inheritance/InheritanceSampleField.cs
// Copyright 2012 rubicon informationstechnologie gmbh
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using InjectionCop.Fragment;

namespace InjectionCop.IntegrationTests.Parser.TypeParsing.TypeParserTests.Inheritance
{
  public class InheritanceSampleField : InheritanceSampleBase
  {
    [Fragment ("InheritanceFragment")]
    protected new string _initialNonFragmentField;

    protected new string _initialFragmentField;

    public InheritanceSampleField ()
        : base ("safe", "safe")
    {
    }

    protected void UnsafeAssignmentOnInheritedField ()
    {
      _fragmentField = UnsafeInheritanceFragmentSource();
    }

    protected void SafeAssignmentOnInheritedFieldWithLiteral ()
    {
      _fragmentField = "safe";
    }

    protected void SafeAssignmentOnInheritedField ()
    {
      _fragmentField = SafeInheritanceFragmentSource();
    }

    protected void SafeAssignmentOnFieldSetByHidingParent ()
    {
      _initialNonFragmentField = SafeInheritanceFragmentSource();
    }

    protected void UnsafeAssignmentOnFieldResetByHidingParent ()
    {
      _initialFragmentField = UnsafeInheritanceFragmentSource();
    }

    protected void SafeCallWithInheritedField ()
    {
      RequiresInheritanceFragment (_fragmentField);
    }

    protected void UnsafeCallWithInheritedField ()
    {
      RequiresInheritanceFragment (_nonFragmentField);
    }

    protected void SafeCallWithFieldSetByHidingParentField ()
    {
      RequiresInheritanceFragment (_initialNonFragmentField);
    }

    protected void UnsafeCallWithFieldSetByHidingParentField ()
    {
      RequiresInheritanceFragment (_initialFragmentField);
    }

    protected void SafeAssignmentOnBaseField ()
    {
      base._initialFragmentField = SafeInheritanceFragmentSource();
    }

    protected void UnsafeAssignmentOnBaseField ()
    {
      base._initialFragmentField = UnsafeInheritanceFragmentSource();
    }
  }
}

File: IntegrationTests.VB/Parser/TypeParsing/TypeParserTests/Inheritance/Field/InheritanceSampleField.vb
' Copyright 2013 rubicon informationstechnologie gmbh
'
' Licensed under the Apache License, Version 2.0 (the "License");
' you may not use this file except in compliance with the License.
' You may obtain a copy of the License at
'
' http://www.apache.org/licenses/LICENSE-2.0
'
' Unless required by applicable law or agreed to in writing, software
' distributed under the License is distributed on an "AS IS" BASIS,
' WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
' See the License for the specific language governing permissions and
' limitations under the License.
Imports InjectionCop.Attributes
Imports System

Namespace Parser.TypeParsing.TypeParserTests.Inheritance.Field
	Public Class InheritanceSampleField
		Inherits InheritanceSampleBase

    <Fragment("InheritanceFragment")>
    Protected Shadows _initialNonFragmentField As String

    Protected Shadows _initialFragmentField As String

		Public Sub New()
      MyBase.New("safe", "safe")
		End Sub

		Protected Sub UnsafeAssignmentOnInheritedField()
			Me._fragmentField = InheritanceSampleBase.UnsafeInheritanceFragmentSource()
		End Sub

		Protected Sub SafeAssignmentOnInheritedFieldWithLiteral()
			Me._fragmentField = "safe"
		End Sub

		Protected Sub SafeAssignmentOnInheritedField()
			Me._fragmentField = MyBase.SafeInheritanceFragmentSource()
		End Sub

		Protected Sub SafeAssignmentOnFieldSetByHidingParent()
			Me._initialNonFragmentField = MyBase.SafeInheritanceFragmentSource()
		End Sub

		Protected Sub UnsafeAssignmentOnFieldResetByHidingParent()
			Me._initialFragmentField = InheritanceSampleBase.UnsafeInheritanceFragmentSource()
		End Sub

		Protected Sub SafeCallWithInheritedField()
			MyBase.RequiresInheritanceFragment(Me._fragmentField)
		End Sub

		Protected Sub UnsafeCallWithInheritedField()
			MyBase.RequiresInheritanceFragment(Me._nonFragmentField)
		End Sub

		Protected Sub SafeCallWithFieldSetByHidingParentField()
			MyBase.RequiresInheritanceFragment(Me._initialNonFragmentField)
		End Sub

		Protected Sub UnsafeCallWithFieldSetByHidingParentField()
			MyBase.RequiresInheritanceFragment(Me._initialFragmentField)
		End Sub

		Protected Sub SafeAssignmentOnBaseField()
      MyBase._initialFragmentField = MyBase.SafeInheritanceFragmentSource()
		End Sub

		Protected Sub UnsafeAssignmentOnBaseField()
      MyBase._initialFragmentField = InheritanceSampleBase.UnsafeInheritanceFragmentSource()
		End Sub

		Protected Sub SafeAssignmentOfBaseField()
      Me._fragmentField = MyBase._initialFragmentField
		End Sub

		Protected Sub UnsafeAssignmentOfBaseField()
      Me._fragmentField = MyBase._initialNonFragmentField
		End Sub

		Public Sub SafeMethodCallUsingBaseField()
      MyBase.RequiresInheritanceFragment(MyBase._initialFragmentField)
		End Sub

		Public Sub UnsafeMethodCallUsingBaseField()
      MyBase.RequiresInheritanceFragment(MyBase._initialNonFragmentField)
		End Sub
	End Class
End Namespace
